---
title: "Introduction to dotplots in R"
author: "Nathan Brouwer"
date: "10/21/2021"
output: html_document
---

# Sequence dotplots in *R*

**By**: Avril Coghlan.  

**Adapted, edited and expanded**: Nathan Brouwer under the Creative Commons 3.0 Attribution License [(CC BY 3.0)](https://creativecommons.org/licenses/by/3.0/).


**NOTE: I've added some new material that is rather terse and lacks explication.**


Good sources of more info: 
 https://omicstutorials.com/interpreting-dot-plot-bioinformatics-with-an-example/ 

http://resources.qiagenbioinformatics.com/manuals/clcgenomicsworkbench/650/Examples_interpretations_dot_plots.html 

As a first step in comparing two protein, RNA or DNA sequences, it is a good idea to make a **dotplot**. A **dotplot** is a graphical method that allows the comparison of two protein or DNA sequences and identify regions of close similarity between them. A dotplot is essentially a two-dimensional matrix (like a grid), which has the sequences of the proteins being compared along the vertical and horizontal axes.

In order to make a simple dotplot to represent of the similarity between two sequences, individual cells in the matrix can be shaded black if residues are identical, so that matching sequence segments appear as runs of diagonal lines across the matrix. Identical proteins will have a line exactly on the main diagonal of the dotplot, that spans across the whole matrix.

For proteins that are not identical, but share regions of similarity, the dotplot will have shorter lines that may be on the **main diagonal**, or off the main diagonal of the matrix. In essence, a dotplot will reveal if there are any regions that are clearly very similar in two protein (or DNA) sequences.

## Preliminaries

```{r, message = F, warning = F}
library(compbio4all)
library(rentrez)
```




## Visualzing two identical sequences

To help build our intuition about dotplots we'll first look at some artificial examples.  First, we'll see what happens when we make a dotplot comparing the alphabet versus itself.  The build-in `LETTERS` object in R contains the alphabet from A to Z.  This is a sequence with no repeats.

```{r}
#LETTERS
seqinr::dotPlot(LETTERS, LETTERS)
```

What we get is a perfect diagonal line.


## Visualizing repeats

TODO: explain this

```{r}
LETTERS.2.times <- c(LETTERS, LETTERS)

seqinr::dotPlot(LETTERS.2.times, 
                LETTERS.2.times)
```



TODO: explain this

```{r}
LETTERS.3.times <- c(LETTERS, LETTERS, LETTERS)
seqinr::dotPlot(LETTERS.3.times, LETTERS.3.times)
```


TODO: explain this

```{r}
seq.repeat <- c("A","C","D","E","F","G","H","I")

seq1<- rep(seq.repeat, 3)

```

Make the dotplot:
```{r}
seqinr::dotPlot(seq1, seq1)

```


## TODO: Add title

TODO: explain this

"invert" means "inversion"
```{r}
LETTERS.3.times.with.invert <- c(LETTERS, rev(LETTERS), LETTERS)

seqinr::dotPlot(LETTERS.3.times.with.invert, LETTERS.3.times.with.invert)
```

## TODO: explain this

TODO: explain this

```{r}
seg1 <- LETTERS[1:8]
seg2 <- LETTERS[9:18]
seg3 <- LETTERS[19:26]


LETTERS.with.transloc <- c(seg1, seg2, seg3)

seqinr::dotPlot(LETTERS.with.transloc, LETTERS.with.transloc)
```


## TODO: add title

TODO: explain this

```{r}
sample(x = LETTERS, size = 26, replace = T)
```


```{r}
letters.rand1 <- sample(x = LETTERS, size = 26, replace = F)
letters.rand2 <- sample(x = LETTERS, size = 26, replace = F)


seqinr::dotPlot(letters.rand1, letters.rand2)
```


## Download sequences

Now we'll make a real dotplot of the [chorismate lyase](https://en.wikipedia.org/wiki/Chorismate_lyase) proteins from two closely related species, *Mycobacterium leprae* and *Mycobacterium ulcerans*.

Note - these are protein sequences so db = "protein"

TODO: briefly summarize these steps

```{r}
# sequence 1: Q9CD83
leprae_fasta <- rentrez::entrez_fetch(db = "protein",
                        id = "Q9CD83",
                         rettype = "fasta")
# sequence 2: OIN17619.1
ulcerans_fasta <- rentrez::entrez_fetch(db ="protein",
                         id = "OIN17619.1",
                         rettype = "fasta")

# add code
leprae_vector   <- compbio4all::fasta_cleaner(leprae_fasta)
ulcerans_vector <- compbio4all::fasta_cleaner(ulcerans_fasta)
```


## TODO: add title



We can create a dotplot for two sequences using the `dotPlot()` function in the `seqinr`  package.

First, let's look at a dotplot created using only a single sequence. This is frequently done to investigate a sequence for the presence of repeats.

(Note - and older version of this exercise stated this kind of anlysis wasn't normally done; this was written last year before I knew of the use of dotplots for investigating sequence repeats.)

```{r}

seqinr::dotPlot(leprae_vector, ulcerans_vector)
```

TODO: briefly explain the main pattern



In the dotplot above, the *M. leprae* sequence is plotted along the x-axis (horizontal axis), and the *M. ulcerans* sequence is plotted along the y-axis (vertical axis). The dotplot displays a dot at points where there is an identical amino acid in the two sequences.

For example, if amino acid 53 in the *M. leprae* sequence is the same amino acid (eg. “W”) as amino acid 70 in the M. ulcerans sequence, then the dotplot will show a dot the position in the plot where x =50 and y =53.

In this case you can see a lot of dots along a diagonal line, which indicates that the two protein sequences contain many identical amino acids at the same (or very similar) positions along their lengths. This is what you would expect, because we know that these two proteins are **homologs** (related proteins) because they share a close evolutionary history.



